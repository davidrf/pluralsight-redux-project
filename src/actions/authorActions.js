import AuthorApi from "../api/mockAuthorApi";
import * as types from "./actionTypes";
import { beginAjaxCall } from "./ajaxStatusActions";

export function loadAuthorsSuccess(authors) {
    return { type: types.LOAD_AUTHORS_SUCCESS, authors };
}

export default function loadAuthors() {
    return function(dispatch) {
        dispatch(beginAjaxCall());
        // Advantages of putting dispatch inside this thunk and NOT in the mock API call (though the latter would certainly centralize this dispatch):
        // Ups the speed of the UI response (we don't have to wait for the API call to finish) -- "optimistic update"
        // Allows us control over which thunks make use of the preloader and which ones don't
        return AuthorApi.getAllAuthors()
        .then(authors => {
            dispatch(loadAuthorsSuccess(authors));
        }).catch(error => {
            throw(error);
        });
    };
}
// this is a thunk -- an action creator that returns a function instead of an action
// loadAuthorsSuccess is an action creator that returns a regular action for when the thunk is successful
// normally, we'd have an action to account for failure as well, but we're not doing that here for the sake of time