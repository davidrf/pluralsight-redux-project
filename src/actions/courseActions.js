import * as types from "./actionTypes";
import courseApi from "../api/mockCourseApi";
import { beginAjaxCall, ajaxCallError } from "./ajaxStatusActions";

export function loadCoursesSuccess(courses) {
    return { type: types.LOAD_COURSES_SUCCESS, courses };
}
// this is an action creator, an optional function that returns an action
// actions have a type and some data to pass along to the reducer

export function updateCourseSuccess(course) {
    return { type: types.UPDATE_COURSE_SUCCESS, course };
}

export function createCourseSuccess(course) {
    return { type: types.CREATE_COURSE_SUCCESS, course };
}

export default function loadCourses() {
    return function(dispatch) {
        dispatch(beginAjaxCall());
        return courseApi.getAllCourses()
        .then(courses => {
            dispatch(loadCoursesSuccess(courses));
        })
        .catch(error => {
            throw(error);
        });
    };
}

export function saveCourse(course) {
    return function(dispatch, getState) {
        // getState allows us to access Redux store directly
        dispatch(beginAjaxCall());
        return courseApi.saveCourse(course)
        .then(savedCourse => {
            course.id ? dispatch(updateCourseSuccess(savedCourse)) : dispatch(createCourseSuccess(savedCourse));
        }).catch(error => {
            dispatch(ajaxCallError(error));
            throw(error);
        });
    };
}

