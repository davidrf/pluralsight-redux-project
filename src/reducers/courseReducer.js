import * as types from "../actions/actionTypes";
import initialState from "./initialState";

export default function courseReducer(state = initialState.courses, action) {
    // state is an array because we want to be able to collect/list all the different courses
    switch(action.type) {
        case types.LOAD_COURSES_SUCCESS:
            return action.courses;
             // takes the data sent through the action
             // creates a new course object with Object.assign({}, action.course)
             // uses spread operator to return a copy of the state with the new data (in line with immutability principle)
        case types.CREATE_COURSE_SUCCESS:
            return [ ...state, Object.assign({}, action.course) ];
            // will return a new array of courses
        case types.UPDATE_COURSE_SUCCESS:
            return [ ...state.filter(course => course.id !== action.course.id), Object.assign({}, action.course) ];
        default: 
            return state;
    }
}