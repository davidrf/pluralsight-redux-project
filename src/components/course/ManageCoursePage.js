import React, { PropTypes } from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import * as courseActions from "../../actions/courseActions";
import * as authorActions from "../../actions/authorActions";
import CourseForm from "./CourseForm";
import toastr from "toastr";
import { withRouter } from 'react-router';

export class ManageCoursePage extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            course: Object.assign({}, this.props.course),
            errors: {},
            saving: false
        };
        this.toastr = toastr;
        this.updateCourseState = this.updateCourseState.bind(this);
        this.saveCourse = this.saveCourse.bind(this);
    }

    componentWillReceiveProps(nextProps) {
        if (this.props.course.id !== nextProps.course.id) {
            // Necessary to populate form when existing course is loaded directly
            // Basically informs React whether or not a change has ACTUALLY occurred
            this.setState({course: Object.assign({}, nextProps.course)});
        }
    }

    updateCourseState(event) {
        const field = event.target.name;
        let course = Object.assign({}, this.state.course);
        course[field] = event.target.value;
        //changes the appearance of the actual field to reflect the user's input
        return this.setState({course: course});
        // return the new state of the course (with the user input)
    }

    saveCourse(event) {
        event.preventDefault();
        this.setState({ saving: true });
        this.props.actions.saveCourse(this.state.course)
            .then(() => this.redirect())
            .catch((error) => {
                this.toastr.error(error);
                this.setState({ saving: false });
            });
        // we can access the actions from our props because we mapped the dispatch to our props below?
    }

    redirect() {
        this.setState({ saving: false });
        this.toastr.success("Course saved!");
        this.props.router.push("/courses");
    }

    render() {
        return (
            <div>
                <CourseForm
                    allAuthors={this.props.authors}
                    onSave={this.saveCourse}
                    onChange={this.updateCourseState}
                    course={this.state.course}
                    errors={this.state.errors}
                    saving={this.state.saving}
                />
            </div>
        );
    }
}

ManageCoursePage.propTypes = {
    course: PropTypes.object.isRequired,
    authors: PropTypes.array.isRequired,
    actions: PropTypes.object.isRequired
};

function getCourseById(courses, id) {
    const course = courses.filter(course => course.id == id);
    if (course.length > 0) return course[0]; //since filter returns an array, get inside and grab the remaining value with bracket notation
    return null;
}

function mapStateToProps(state, ownProps) {
    const courseId = ownProps.params.id; //from the path "/course/:id"
    let course = {id: " ", watchHref: " ", title: " ", authorId: " ", length: " ", category: " "};

    if (courseId && state.courses.length > 0) {
        // state.courses is coming from our store's state
        course = getCourseById(state.courses, courseId);
    }

    const authorsFormattedForDropdown = state.authors.map(author => {
        return {
            value: author.id,
            text: `${author.firstName} ${author.lastName}`
        };
    });
    return {
        course: course,
        authors: authorsFormattedForDropdown
    };
}

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators(courseActions, dispatch)
    };
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(ManageCoursePage));
