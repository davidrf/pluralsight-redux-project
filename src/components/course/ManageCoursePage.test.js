import { connect, Provider } from "react-redux";
import * as actions from "../../actions/courseActions"
import { createMockStore } from "redux-test-utils";
import React from "react";
import nock from 'nock'
import { shallowWithStore } from "enzyme-redux";
import { mount } from "enzyme"
import thunk from 'redux-thunk'
import renderer from 'react-test-renderer'
import ManageCoursePage from "./ManageCoursePage";
import courseReducer from "../../reducers/courseReducer";

describe("ManageCoursePage", () => {

    afterEach(() => {
        nock.cleanAll()
    });

    it("should render properly", () => {
        const firstCourse =  {
            id: "architecture",
            title: "Architecting Applications for the Real World",
            watchHref: "http://www.pluralsight.com/courses/architecting-applications-dotnet",
            authorId: "cory-house",
            length: "2:52",
            category: "Software Architecture"
        };
        const expectedState = { authors: ["Cory House"], courses: [firstCourse], numAjaxCallsInProgress: 0};
        const mapStateToProps = (state) => ({
            state,
          });
        const ConnectedManageCoursePage = connect(mapStateToProps)(ManageCoursePage);

        const wrapper = renderer.create(
          <Provider store={createMockStore(expectedState)}>
              <ConnectedManageCoursePage params={"architecture"}/>
              {/* need to add in "ownProps" that don't come from state */}
          </Provider>,
        ).toJSON();
        expect(wrapper).toMatchSnapshot();
    });

    it("should have the correct default props from state", () => {
      const expectedState =  { authors: [], courses: [], numAjaxCallsInProgress: 0};
      const mapStateToProps = (state) => ({
        state,
      });
      const ConnectedManageCoursePage = connect(mapStateToProps)(ManageCoursePage);
      const component = shallowWithStore(<ConnectedManageCoursePage />, createMockStore(expectedState));
      expect(component.props().state).toBe(expectedState);
    });

    it("should dispatch the correct actions", () => {
        const firstCourse =  {
            id: "architecture",
            title: "Architecting Applications for the Real World",
            watchHref: "http://www.pluralsight.com/courses/architecting-applications-dotnet",
            authorId: "cory-house",
            length: "2:52",
            category: "Software Architecture"
        };
        const expectedState = { authors: ["Cory House"], courses: [firstCourse], numAjaxCallsInProgress: 0};
        const mapStateToProps = (state) => ({
            state,
          });

        const secondCourse = {
            id: "career-reboot-for-developer-mind",
            title: "Becoming an Outlier: Reprogramming the Developer Mind",
            watchHref: "http://www.pluralsight.com/courses/career-reboot-for-developer-mind",
            authorId: "cory-house",
            length: "2:30",
            category: "Career"
        }

        const action = {
            type: 'CREATE_COURSE',
            data: secondCourse
        };

        const mapDispatchToProps = (dispatch) => ({
            saveCourse(course) {
                dispatch(action);
            },
        });

        const store = createMockStore(expectedState);
        const ConnectedManageCoursePage = connect(mapStateToProps, mapDispatchToProps)(ManageCoursePage);
        const component = shallowWithStore(<ConnectedManageCoursePage />, store);

        // return store.dispatch(actions.saveCourse(secondCourse)).then(() => {
        //     console.log(store.getActions());
        // })
        // console.log(store.getActions());
        // expect(store.isActionDispatched(action)).toBe(true);
    })

    // it("should dispatch the correct ASYNC actions, round TWO", () => {

    // })
  });
