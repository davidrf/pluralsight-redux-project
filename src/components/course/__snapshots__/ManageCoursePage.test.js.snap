// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageCoursePage should render properly 1`] = `
<div>
  <form>
    <h1>
      Manage Course
    </h1>
    <div
      className="form-group"
    >
      <label
        htmlFor="title"
      >
        Title
      </label>
      <div
        className="field"
      >
        <input
          className="form-control"
          name="title"
          onChange={[Function]}
          placeholder={undefined}
          type="text"
          value=" "
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="authorId"
      >
        Author
      </label>
      <div
        className="field"
      >
        <select
          className="form-control"
          name="authorId"
          onChange={[Function]}
          value=" "
        >
          <option
            value=" "
          >
            Select Author
          </option>
          <option
            value={undefined}
          >
            undefined undefined
          </option>
        </select>
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="category"
      >
        Category
      </label>
      <div
        className="field"
      >
        <input
          className="form-control"
          name="category"
          onChange={[Function]}
          placeholder={undefined}
          type="text"
          value=" "
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="length"
      >
        Length
      </label>
      <div
        className="field"
      >
        <input
          className="form-control"
          name="length"
          onChange={[Function]}
          placeholder={undefined}
          type="text"
          value=" "
        />
      </div>
    </div>
    <input
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="submit"
      value="Save"
    />
  </form>
</div>
`;
